GeoShapeType.compare must return 0 for equal polygons

BaurzhanSakhariev opened the pull request:
Fix equality check for geo shape. New tests fail on master.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/GeoShapeIntegrationTest.java:
```diff
@@ -251,4 +250,32 @@ public void test_geo_shape_can_be_used_as_generated_column() throws Exception {
         );
     }
+
+    @Test
+    public void test_polygon_can_be_used_as_generated_column() {
+        // Dynamic nested array is currently not supported, declaring coordinates array upfront.
+        execute("""
+            create table tbl
+            (
+                o object as (
+                    o1 object as (coordinates ARRAY(ARRAY(ARRAY(DOUBLE))), type string),
+                    geo as o['o1']::geo_shape
+               )
+            )
+            """
+        );
+        List<List<List<Double>>> coords = List.of(
+            List.of(List.of(0.0, 0.0), List.of(1.0, 0.0), List.of(1.0, 1.0), List.of(0.0, 0.0))
+        );
+        execute(
+            "insert into tbl (o) values (?)",
+            new Object[] {
+                Map.of("o1", Map.of("coordinates", coords, "type", "Polygon"))
+            }
+        );
+        execute("refresh table tbl");
+        assertThat(execute("select o['geo'] from tbl")).hasRows(
+            "{coordinates=[[[0.0, 0.0], [1.0, 0.0], [1.0, 1.0], [0.0, 0.0]]], type=Polygon}"
+        );
+    }
```
I decided to add a test because we already had a case of fixing one part of generated column aspect and missing some scenario due to refactoring or something else. See https://github.com/crate/crate/pull/16378#issuecomment-2247377927

BaurzhanSakhariev commented on server/src/test/java/io/crate/types/GeoShapeTypeTest.java:
```diff
@@ -119,16 +119,6 @@ private static Map<String, Object> parse(String json) {
 
 
-    @Test
-    public void testCompareValueTo() throws Exception {
-        Map<String, Object> val1 = type.implicitCast("POLYGON ( (0 0, 20 0, 20 20, 0 20, 0 0 ))");
-        Map<String, Object> val2 = type.implicitCast("POINT (10 10)");
-
-        assertThat(type.compare(val1, val2)).isEqualTo(1);
-        assertThat(type.compare(val2, val1)).isEqualTo(-1);
-        assertThat(type.compare(val2, val2)).isEqualTo(0);
-    }
-
```
Order is undefined, we even have a test `testOrderByGeoShape`

BaurzhanSakhariev commented on libs/shared/src/main/java/io/crate/common/collections/Maps.java:
```diff
@@ -51,6 +52,40 @@ public static <K, V> Map<K, V> concat(Map<K, V> m1, Map<K, V> m2) {
+                    if (!(value2 == null && m2.containsKey(key)))
+                        return false;
+                } else {
+                    // Compares arrays by values.
+                    if (!Objects.deepEquals(value, value2)) {
+                        return false;
+                    }
```
Please note that this still requires arguments be the arrays of the same type. 
Ie we cannot compare `double[], Double[]` it will fall back to identity equals.

That's why I changed `test_cast_wkt_point_string_array_to_geo_shape_array` to have primitive in the assertion - because in the real execution we cannot have different types and `Double[]` usage in the assertion is rather artifical.

Another option could be to get back GeoShape construction and check equality like
`shape1.within(shape2) and shape2.within(shape1)`- but I don't think it's ay better and we even have a TODo to compare without building shapes.

seut commented on docs/appendices/release-notes/5.8.5.rst:
```diff
@@ -58,3 +58,6 @@ Fixes
 - Fixed a regression introduced in 5.8.4 leading to a rejection of writes into
   a column of the ``OBJECT(IGNORED)`` type if it had an array sub-column with
   mixed types.
+
+- Fixed an issue that caused write operations to fail if a table contained
+  generated ``geo_shape`` column with ``Polygon`` or ``MultiPolygon`` type.
```
I think we should be more specific. Additionally, this is not only affecting `GeoShape` values containing `Polygon` or `MultiPolygon` as the new `test_check_equality` unit test will also fail for `LineString` values.

Not sure if my suggestion is correct:

```suggestion
  mixed types.

- Fixed an issue that caused write operations to fail if a table contains a
  generated ``geo_shape`` column and the user provides a correct value for this generated column.
```

BaurzhanSakhariev replied:
even "value is provided" case fails only for Polygons/MultiPolygons and LineString/MultiLineString (thanks for catching!).

I combined your suggestion and added enumeration of problematic shapes.

seut commented on libs/shared/src/main/java/io/crate/common/collections/Maps.java:
```diff
@@ -51,6 +52,40 @@ public static <K, V> Map<K, V> concat(Map<K, V> m1, Map<K, V> m2) {
+     * Similar to {@link java.util.AbstractMap#equals(Object)}}
+     * but compares nested arrays by values instead of relying on default Array implementation comparing by reference.
+     */
+    public static boolean equals(Map<?, ?> m1, Map<?, ?> m2) {
```
I don't see any `Array` specific logic here. Why not use the existing `MapComparator.compareMaps()` instead? All test pass when using this.

seut commented on server/src/main/java/io/crate/types/GeoShapeType.java:
```diff
@@ -124,14 +124,13 @@ public Map<String, Object> sanitizeValue(Object value) {
-            case CONTAINS -> 1;
-            default -> Double.compare(shape1.getArea(JtsSpatialContext.GEO), shape2.getArea(JtsSpatialContext.GEO));
-        };
+        if (Maps.equals(val1, val2)) {
+            return 0;
+        } else {
+            // Order is undefined for GEO SHAPES, returning arbitrary value.
+            // TODO: Add a method to the DataType indicating comparison support and use it instead of SemanticSortValidator.SUPPORTED_TYPES.
+            return 1;
+        }
```
```suggestion
        return MapComparator.compareMaps(val1, val2);
```

BaurzhanSakhariev replied:
Thanks, I was not aware that we have it already

BaurzhanSakhariev replied:
> I don't see any Array specific logic here

`Objects.deepEquals` has that logic.

On MapsComparator: Tests indeed are passing but if I revert change done in `test_cast_wkt_point_string_array_to_geo_shape_array` test fails with 

`java.lang.IllegalArgumentException: Cannot detect the type of the value: [D@32ab68`
(it cannot work with a primitive).

I think it's still fine as we don't expect primitives and it's only artificially coming from the assertion?
See also https://github.com/crate/crate/pull/16761#discussion_r1795478418

seut replied:
Our type system doesn't support primitive types as values, such we must ensure to always use non-primitive values when we want to detect or sanitize them. This can be fixed by parsing GeoShape coordinates as `Double[]` instead of `double[]`. I've pushed a commit demonstrating this suggestion.

> Fix DataType.guessType to deal with Double[] (it's not hitting case Object[] array).

This won't help/work as the parsed GeoShape coordinates we want to compare with will still return `double[]`

 > Avoid type guessing in the MapsComparator.

BaurzhanSakhariev replied:
Thanks for the suggestion. While fixing a test with size estimations (test_bytes_estimate has some entries with a bigger expected value due to Double Object overhead).

However, some entries got lower which was surprusing. I investigated a bit, looks like we are not hitting primitive array based logic introduced in 
https://github.com/crate/crate/commit/90f19dbe54241d0dfb21ed44b859831c5d2bb04b, I will fix it as well

BaurzhanSakhariev replied:
Pushed an adjustment of geo shape accounting https://github.com/crate/crate/pull/16761/commits/ec87d239423a8830a7272e49a593406f06db2897

seut commented:
@mfussenegger What do you think about the change of using `Double[]` instead of `double[]` here, any objections or better ideas?

BaurzhanSakhariev commented:
Retest this please 
https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/17215/testReport/junit/io.crate.gcs/GCSBlobContainerRetriesTests/testReadBlobWithReadTimeouts/

mfussenegger commented:
> @mfussenegger What do you think about the change of using `Double[]` instead of `double[]` here, any objections or better ideas?

Not a fan as it ruins data locality and increases the memory footprint.
Long term I'd rather move to a more special representation (similar to https://github.com/crate/crate/pull/16722, but there are quite a few design question - e.g. if we could go with a spatial4j representation directly - otherwise we have JSON (pgwire/http) -> GeoShape (our structure) -> spatial4j/lucene-geometry (depending on operations)  conversions)


I wouldn't try to handle this with a generic map comparison.
We know the shape/structure of GeoJSON, so we could make the comparison more concrete on the `type` and `coordinates` properties.

Handling both `List<Double>` and `double[]` with up to 4 dimensions (multi polygon) will still be annoying but should be doable?


> Our type system doesn't support primitive types as values, such we must ensure to always use non-primitive values when we want to detect or sanitize them

While it is true that a `object (dynamic|strict)` type can only have values supported by other types, the other types are allowed to use representations as they see fit. E.g. `float_vector` uses `float[]`.

BaurzhanSakhariev commented:
>I wouldn't try to handle this with a generic map comparison.

>Handling both List<Double> and double[] with up to 4 dimensions (multi polygon) will still be annoying but should be doable?

I will check it out

BaurzhanSakhariev commented:
> We know the shape/structure of GeoJSON, so we could make the comparison more concrete on the type and coordinates properties.

I started with this, geometries collection is a bit tricky. I thought that we cannot iterate over `shape1.geometries` and `shape2.geometries` and `compare(geom1[i], geom2[i])` as they can be equal but have multiple geometries of the same type defined in arbitrary order. ie 
```
collection1 
  point1, point2
collection2 
 point2, point1
```

But after checking `GeometryCollectionBuilder.equals` looks like order is important, so it simplifies things - will follow this logic

BaurzhanSakhariev commented:
I reverted double[] -> Double[] change and added a method to compare maps based on **valid** geo shape specific properties.

I assumed that compare already takes valid values - but can also add existing `validateGeoJson` on top of the compare.

mfussenegger commented on server/src/main/java/io/crate/geo/GeoJSONUtils.java:
```diff
@@ -155,6 +180,54 @@ public static Object map2LuceneShape(Map<String, Object> geoJSONMap) {
+            Iterator<?> it1 = iterator(geometries1);
+            Iterator<?> it2 = iterator(geometries2);
+            result = 0;
+            while (it1.hasNext()) {
+                assert it2.hasNext() : "Collections must have the same size";
+                Object value1 = it1.next();
+                Object value2 = it2.next();
+                assert value1 instanceof Map : "Shapes must be valid";
+                assert value2 instanceof Map : "Shapes must be valid";
+                result += compare((Map<String, Object>) value1, (Map<String, Object>) value2);
+            }
+        } else {
+            Object coordinates1 = val1.get(COORDINATES_FIELD);
+            Object coordinates2 = val2.get(COORDINATES_FIELD);
+            assert coordinates1 != null && coordinates2 != null : "Shapes must be valid";
+            Iterator<?> it1 = iterator(coordinates1);
+            Iterator<?> it2 = iterator(coordinates2);
+            result = 0;
+            while (it1.hasNext()) {
+                assert it2.hasNext() : "Collections must have the same size";
+                Object value1 = it1.next();
+                Object value2 = it2.next();
+                result += value1.equals(value2) ? 0 : 1;
+            }
+        }
+        return result > 0 ? 1 : 0;
+    }
```
This could short circuit on the first non-equal value.

mfussenegger commented on server/src/main/java/io/crate/geo/GeoJSONUtils.java:
```diff
@@ -110,6 +110,31 @@ private static void forEach(Object arrayOrCollection, Consumer<Object> consumer)
     }
 
+    private static Iterator<?> iterator(Object arrayOrCollection) {
+        if (arrayOrCollection.getClass().isArray()) {
+            return new Iterator<>() {
+                int arrayLength = Array.getLength(arrayOrCollection);
+                int current = 0;
+
+                @Override
+                public boolean hasNext() {
+                    return current < arrayLength;
+                }
+
+                @Override
+                public Object next() {
+                    int c = current;
+                    current++;
+                    return Array.get(arrayOrCollection, c);
+                }
+            };
+        } else if (arrayOrCollection instanceof Collection<?> collection) {
+            return collection.iterator();
+        } else {
+            throw new AssertionError("argument is neither an array nor a collection");
+        }
```
Not nice to do boxing for the comparison if the input format is `double[]` on both sides.
I think this would be worth special casing.

BaurzhanSakhariev commented:
Found a bug (and fixed it), see https://github.com/crate/crate/pull/16761/commits/4fe4f2fa8d3467f1874f8d6e4824922adaf82112

mfussenegger commented on server/src/test/java/io/crate/types/GeoShapeTypeTest.java:
```diff
@@ -223,5 +213,53 @@ public void test_bytes_estimate() throws Exception {
         assertThat(type.valueBytes(normalize.apply(geoJson.get(7)))).isEqualTo(880);
     }
+
+    @Test
+    public void test_check_equality() throws Exception {
+        for (Map<String, Object> json: geoJson) {
+            assertThat(type.compare(json, json)).isEqualTo(0);
+        }
+    }
```
```suggestion
    @Test
    public void test_check_equality() throws Exception {
        UnaryOperator<Map<String, Object>> normalize = geoJson -> {
            return type.implicitCast(GeoJSONUtils.map2Shape(geoJson));
        };
        for (Map<String, Object> json: geoJson) {
            assertThat(type.compare(json, json)).isEqualTo(0);
            assertThat(type.compare(normalize.apply(json), json)).isEqualTo(0);
            assertThat(type.compare(normalize.apply(json), normalize.apply(json))).isEqualTo(0);
        }
    }
```

Can you check with something like this? Looks like this currently fails but I think it should still pass

BaurzhanSakhariev replied:
Looks like `GeoJSONUtils.map2Shape(geoJson)` can trip order of coordinates 
and it breaks `i-th with i-th` element comparison.

BaurzhanSakhariev replied:
I will go with another approach

BaurzhanSakhariev replied:
I pushed another approach (as discussed with @seut) we will do topological equality for shapes but for geometry collections order must be the same. 

Since normalize here might reshuffle geometry collection (after converting to string and parsing it back), I instead added 2 dedicated tests - one for geometry with same but shuffled shapes and one for topological equality (also added a doc clarification/test for that)

mfussenegger commented on server/src/test/java/io/crate/types/GeoShapeTypeTest.java:
```diff
@@ -223,5 +213,53 @@ public void test_bytes_estimate() throws Exception {
+
+    @Test
+    public void test_check_equality_of_collections_of_different_sizes() throws Exception {
+        Map<String, Object> collection1 = parse("{ \"type\": \"GeometryCollection\",\n" +
+            "    \"geometries\": [\n" +
+            "      { \"type\": \"Point\",\n" +
+            "        \"coordinates\": [100.0, 0.0]\n" +
+            "        },\n" +
+            "      { \"type\": \"LineString\",\n" +
+            "        \"coordinates\": [ [101.0, 0.0], [102.0, 1.0] ]\n" +
+            "        }\n" +
+            "    ]\n" +
+            "  }");
+        Map<String, Object> collection2 = parse("{ \"type\": \"GeometryCollection\",\n" +
+            "    \"geometries\": [\n" +
+            "      { \"type\": \"Point\",\n" +
+            "        \"coordinates\": [100.0, 0.0]\n" +
+            "        },\n" +
+            "      { \"type\": \"LineString\",\n" +
+            "        \"coordinates\": [ [101.0, 0.0], [102.0, 1.0] ]\n" +
+            "        },\n" +
+            "      { \"type\": \"LineString\",\n" +
+            "        \"coordinates\": [ [101.0, 0.0], [102.0, 1.0] ]\n" +
+            "        }\n" +
+            "    ]\n" +
+            "  }");
```
Could use multiline strings to make this more readable:

```suggestion
        Map<String, Object> collection1 = parse(
            """
            {
                "type": "GeometryCollection",
                "geometries": [
                    {
                        "type": "Point",
                        "coordinates": [100.0, 0.0]
                    },
                    {
                        "type": "LineString",
                        "coordinates": [[101.0, 0.0], [102.0, 1.0]]
                    }
                ]
            }
            """
        );
        Map<String, Object> collection2 = parse(
            """
            {
                "type": "GeometryCollection",
                "geometries": [
                    {
                        "type": "Point",
                        "coordinates": [100.0, 0.0]
                    },
                    {
                        "type": "LineString",
                        "coordinates": [[101.0, 0.0], [102.0, 1.0]]
                    },
                    {
                        "type": "LineString",
                        "coordinates": [[101.0, 0.0], [102.0, 1.0]]
                    }
                ]
            }
            """
        );
```

mfussenegger commented on server/src/main/java/io/crate/geo/GeoJSONUtils.java:
```diff
@@ -155,6 +181,81 @@ public static Object map2LuceneShape(Map<String, Object> geoJSONMap) {
+        int result;
+        if (GEOMETRY_COLLECTION.equals(type1)) {
+            // Both shapes are valid geometries collections.
+            Object geometries1 = val1.get(GeoJSONUtils.GEOMETRIES_FIELD);
+            Object geometries2 = val2.get(GeoJSONUtils.GEOMETRIES_FIELD);
+            Iterator<?> it1 = iterator(geometries1);
+            Iterator<?> it2 = iterator(geometries2);
```
Afaik geometries can't be an array but must be a collection, so could use assert/cast here instead of going through `iterator`

mfussenegger commented on server/src/main/java/io/crate/geo/GeoJSONUtils.java:
```diff
@@ -155,6 +181,81 @@ public static Object map2LuceneShape(Map<String, Object> geoJSONMap) {
+        } else {
+            Object coordinates1 = val1.get(COORDINATES_FIELD);
+            Object coordinates2 = val2.get(COORDINATES_FIELD);
+            if (coordinates1 instanceof double[] arr1 && coordinates2 instanceof double[] arr2) {
+                // Not using compare as it can return -1, geo shape compare returns either 0 or 1.
+                return Arrays.equals(arr1, arr2) ? 0 : 1;
+            }
```
What about linestring, polygon, multipolygon which are `double[][]`, `double[][][]` and `double[][][][]` ?

BaurzhanSakhariev replied:
I will add those cases, thanks. Currently investigating why we end up with Lists here. 

`ShardUpsertRequest.Item` sometimes has primitives but at some point it has Lists, figuring out where this conversion happens.

BaurzhanSakhariev replied:
Ok, it's GeoShapeType.readValueFrom -> readGenericValue -> readArray. Turns double[] into List.

This is not related to this PR, ie I will add extra cases as we can have them when not streaming and array-> List should be handled separately.

BaurzhanSakhariev commented:
This approach doesn't work, see https://github.com/crate/crate/pull/16761#discussion_r1802968625

BaurzhanSakhariev commented:
Hi @seut, can you please have another look?

I switched to another approach of topological equality. 

See also https://postgis.net/docs/ST_Equals.html  - I based our docs update on it, cc @henrikingo.

Regarding the tests suggested by @mfussenegger and that actually caught this case: I added dedicated tests instead of normalizing/convering back and forth since it can affect order of shapes in GeometryCollection (see https://github.com/crate/crate/pull/16761#discussion_r1805048097).

henrikingo commented:
Just noting I came back to review after the fact, I think the doc is clearer now and the link provides even more bakground. Thanks!
